---
title: "trf_output"
author: "Hope"
date: "15/02/2023"
output: html_document
---

```{r}
library('rjson')
library('spgs')
```

```{r}
trf_output = fromJSON(file = "TRF_find_motif/filtered_trf_output_modified.txt") %>% as.data.frame() %>% gather() 
trf_output = trf_output %>% mutate (key = case_when(key == "X16PTEL06"~ "16PTEL06",
                                       key == "X18QTEL11"~ "18QTEL11",
                                       key == "X2QTEL47" ~ "2QTEL47",
                                       key == "X4PTEL04"~ "4PTEL04",
                                       key == "MFD424.TTTA003"~ "MFD424-TTTA003",
                                       key == "MFD442.GTTT002" ~ "MFD442-GTTT002",
                                       key == "MFD455.AAT052"~"MFD455-AAT052",
                                       TRUE ~ key))
```
`

```{r}
fasta = read_excel("UCSC_insilicoPCR/482_loci_with_primers_UCSC_remove_fix_alt.xlsx") %>% select(-...1)
fasta$coordinate = gsub("\r\n", " ", fasta$coordinate)
fasta$coordinate = str_trim(fasta$coordinate, side = "both")
fasta = fasta %>% filter(coordinate != "no match")%>% separate(coordinate, c("coordinates", "PCR_length", "forward_primer", "reverse_primer", "PCR_sequence_1", "PCR_sequence_2", "PCR_sequence_3", "PCR_sequence_4", "PCR_sequence_5", "PCR_sequence_6", "PCR_sequence_7", "PCR_sequence_8"), sep = " ")

fasta = fasta %>% mutate(PCR_sequence_2 = replace_na(PCR_sequence_2, ""),
                  PCR_sequence_3 = replace_na(PCR_sequence_3, ""),
                  PCR_sequence_4 = replace_na(PCR_sequence_4, ""),
                  PCR_sequence_5 = replace_na(PCR_sequence_5, ""),
                  PCR_sequence_6 = replace_na(PCR_sequence_6, ""),
                  PCR_sequence_7 = replace_na(PCR_sequence_7, ""),
                  PCR_sequence_8 = replace_na(PCR_sequence_8, "")
  
)

fasta$PCR_sequence = paste(fasta$PCR_sequence_1, fasta$PCR_sequence_2, fasta$PCR_sequence_3, fasta$PCR_sequence_4, fasta$PCR_sequence_5, fasta$PCR_sequence_6, fasta$PCR_sequence_7, fasta$PCR_sequence_8)

fasta$PCR_sequence = gsub(" ", "", fasta$PCR_sequence)

fasta = fasta %>% select(-PCR_sequence_1, -PCR_sequence_2, - PCR_sequence_3, -PCR_sequence_4, -PCR_sequence_5, -PCR_sequence_6, -PCR_sequence_7, -PCR_sequence_8, -forward_primer, -reverse_primer)

fasta$PCR_length = gsub("bp", "", fasta$PCR_length)
```

```{r}
fasta = merge(trf_output, fasta, all.y = TRUE, by.x = "key", by.y = "alternateName")
fasta = fasta %>% separate(value, c("trf_start_index", "trf_end_index", "trf_period", "trf_num_repeats", "trf_size_consensus", "trf_percent_match", "trf_percent_indels", "trf_alignment", "trf_percent_composition_1", "trf_percent_composition_2", "trf_percent_composition_3", "trf_percent_composition_4", "trf_entropy", "trf_motif", "trf_repeat_sequence"), sep = " ")
fasta= fasta %>% mutate(sense = case_when(
  grepl("-", coordinates)~"antisense",
  grepl("+", coordinates)~ "sense",
  TRUE ~ "nosense"
))
fasta = fasta %>% separate(coordinates, c("chr", "coordinates"), sep = ":") %>% separate(coordinates, c("start_coordinate_PCR", "end_coordinate_PCR"), sep = "[+|-]") 

fasta$chr = gsub(">", "", fasta$chr)

fasta$trf_start_index = as.numeric(fasta$trf_start_index)
fasta$trf_end_index = as.numeric(fasta$trf_end_index)
fasta$end_coordinate_PCR = as.numeric(fasta$end_coordinate_PCR)
fasta$start_coordinate_PCR = as.numeric(fasta$start_coordinate_PCR)

fasta = fasta %>% mutate(catalog_start_coordinate_0_based = case_when(
  sense == "sense" ~ start_coordinate_PCR + trf_start_index-2, 
  sense== "antisense" ~ end_coordinate_PCR - trf_end_index
),
  catalog_end_coordinate = case_when(
sense == "sense"  ~start_coordinate_PCR + trf_end_index -1,
sense == "antisense" ~ end_coordinate_PCR - trf_start_index+1

  )
)

fasta = fasta %>% mutate ( trf_repeat_sequence = case_when(
                  sense == "antisense"~ reverseComplement(trf_repeat_sequence),
                  TRUE ~ trf_repeat_sequence)  
                  )

## creation of trf input to redefine motifs on antisense strand
#trf_input = fasta %>% filter(sense == "antisense") %>% select(chr,catalog_start_coordinate_0_based, catalog_end_coordinate, trf_repeat_sequence)
#trf_input$name = paste(">",trf_input$chr,":" ,trf_input$catalog_start_coordinate_0_based, "-", trf_input$catalog_end_coordinate, sep = "")
#trf_input = trf_input %>% select(name, trf_repeat_sequence)
#write.table(trf_input, file = "trf_input_antisense.txt", sep = "\n", row.names = FALSE, quote = FALSE)

fasta = fasta %>% mutate(catalog_start_coordinate_1_based = catalog_start_coordinate_0_based+1)
```

```{r}
trf_output_antisense = read.csv("redefining_motifs_on_antisense/trf_output_antisense.txt") %>% separate(sequence, c("chr","coordinates" ), sep = ":") %>% separate(coordinates, c("catalog_start_coordinate_0_based", "catalog_end_coordinate"), sep = "-") %>% rename("updated_motif" = "motif")
```


```{r}
fasta = merge(fasta, trf_output_antisense, by = c("chr", "catalog_start_coordinate_0_based", "catalog_end_coordinate"), all.x = TRUE)
```

```{r}
fasta = fasta %>% mutate(trf_motif = case_when(sense == "antisense"~ updated_motif,
                                       TRUE ~trf_motif))
fasta = fasta %>% select(-updated_motif)
fasta %>% filter(catalog_start_coordinate_0_based == 6404707)
```

```{r}

## gangstr catalog creation
gangstr_catalog = fasta %>% select(chr,catalog_start_coordinate_1_based, catalog_end_coordinate, trf_motif) %>% mutate(motif_len= nchar(trf_motif)) %>% select(chr, catalog_start_coordinate_1_based, catalog_end_coordinate, motif_len, trf_motif)

write.table(gangstr_catalog, file = "catalogs/gangstr_catalog_hg38_backbone_1_based.bed", sep = "\t", row.names = FALSE, quote = FALSE)
```

```{r}
capillary_genotypes = read.csv("capillary_genotypes_with_locusNames_rosenberg2005.csv") %>% select(-X,-locus_id, -population_id, - geographic_population, -predefined_region)

fasta_with_capillary = merge(fasta, capillary_genotypes, by.x = "key", by.y = "alternateName", all.x = TRUE) %>% rename("PCR_1" = genotype_1, "PCR_2" = genotype_2)

fasta_with_capillary$PCR_1 = na_if(fasta_with_capillary$PCR_1, -9)
fasta_with_capillary$PCR_2 = na_if(fasta_with_capillary$PCR_2, -9)
fasta_with_capillary$trf_num_repeats = as.numeric(fasta_with_capillary$trf_num_repeats)
fasta_with_capillary$PCR_length = as.numeric(fasta_with_capillary$PCR_length)
```

```{r}
eh = read.csv("str_sensitivity-analysis_data_frames_untrimmed_coordinates_hg38_backbone_1_based_hg38_backbone_untrimmed_1_based_eh.csv") %>% select(sample_id, chr, start, end, e_allele_1, e_alelle_2) %>% rename("catalog_start_coordinate_1_based" = start) %>% mutate(catalog_start_coordinate_0_based = as.numeric(catalog_start_coordinate_1_based)-1)

gangstr = read.delim("str_sensitivity-analysis_data_frames_untrimmed_coordinates_hg38_backbone_1_based_hg38_backbone_untrimmed_1_based_gangstr.tsv", sep = "\t") %>% select(sample_id, chr, start, g_allele_1, g_allele_2)%>% rename("catalog_start_coordinate_1_based" = start) %>% mutate(catalog_start_coordinate_0_based = as.numeric(catalog_start_coordinate_1_based)-1)

eh_gangstr = merge(eh, gangstr, by = c("sample_id", "chr", "catalog_start_coordinate_0_based"), all.x = TRUE) %>% rename("cpg_id" = "sample_id")

fasta_with_capillary_eh_gangstr = merge(eh_gangstr, fasta_with_capillary, by = c("cpg_id", "chr", "catalog_start_coordinate_0_based"), all.x = TRUE)

fasta_with_capillary_eh_gangstr = fasta_with_capillary_eh_gangstr %>% mutate(trf_num_repeats = (end -catalog_start_coordinate_0_based)/as.numeric(trf_period))

fasta_with_capillary_eh_gangstr %>% select(trf_num_repeats)

```

```{r}
fasta_with_capillary_eh_gangstr = fasta_with_capillary_eh_gangstr %>% mutate(capillary_allele_1 = floor(trf_num_repeats) + (PCR_1 -PCR_length)/nchar(trf_motif),
                               capillary_allele_2 = floor(trf_num_repeats) + (PCR_2 -PCR_length)/nchar(trf_motif) )

fasta_with_capillary_eh_gangstr = fasta_with_capillary_eh_gangstr %>% mutate(e_alleles = e_allele_1+e_alelle_2, 
                 g_alleles = g_allele_2+g_allele_1,
                 capillary_alleles = floor(capillary_allele_1) + floor(capillary_allele_2)) #summed diploid genotypes 
```

```{r}
fasta_with_capillary_eh_gangstr =fasta_with_capillary_eh_gangstr %>% filter(!is.na(capillary_alleles))
fasta_with_capillary_eh_gangstr %>% filter(e_alleles == g_alleles) %>% count()/ (fasta_with_capillary_eh_gangstr %>% count())
(fasta_with_capillary_eh_gangstr %>% filter(e_alleles== capillary_alleles) %>%count())/ (fasta_with_capillary_eh_gangstr %>% count())
(fasta_with_capillary_eh_gangstr %>% filter(g_alleles== capillary_alleles) %>%count())/ (fasta_with_capillary_eh_gangstr %>% count())
```

```{r}
pemberton = read.table("Pemberton_AdditionalFile1_11242009.txt", sep = "\t", header = TRUE) 
pemberton = merge(fasta, pemberton, by.x = "key", by.y = "alternateName", all.x = TRUE)
pemberton = pemberton %>% separate(repeatStructureInRefSeq, c("motif", "pemberton_repeat_number"), sep = "\\)")
pemberton %>% select(key, trf_num_repeats, pemberton_repeat_number) %>% rename("TRF_num_repeats_hg38" = trf_num_repeats, "Pemberton_2009_num_repeats" = pemberton_repeat_number)
```

